### ISSUE TITLE
How to speed the overlaps filter process?

### ISSUE THREAD

**Issue by RyanBio** [marbl/canu#602]
Hi, developers
My canu version is v1.5 +185 changes. When set useGrid=true, Canu have both parallel and serial jobs, but serial job like filterCorrectionOverlaps takes sevaral days. The cluster is free when running this process, speed it can largely reduce time cost.

I notice the filterCorrectionOverlaps filter overlaps step by step, the log file:
```
        1 -     11 overlaps -     11 scored -      0 filtered -   11 saved (no filtering)
        2 -      9 overlaps -      9 scored -      0 filtered -    9 saved (no filtering)
        3 -   1689 overlaps -   1689 scored -   1629 filtered -   60 saved (length * erate cutoff 3899.08)
        4 -    177 overlaps -    177 scored -    117 filtered -   60 saved (length * erate cutoff 1278.32)
        5 -     15 overlaps -     15 scored -      0 filtered -   15 saved (no filtering)
        6 -     12 overlaps -     12 scored -      0 filtered -   12 saved (no filtering)
        7 -   2626 overlaps -   2626 scored -   2565 filtered -   61 saved (length * erate cutoff 3905.74)
        8 -   6250 overlaps -   6250 scored -   6190 filtered -   60 saved (length * erate cutoff 1081.23)
        9 -     26 overlaps -     26 scored -      0 filtered -   26 saved (no filtering)
       10 -    880 overlaps -    880 scored -    820 filtered -   60 saved (length * erate cutoff 306.78)
       11 -   2851 overlaps -   2851 scored -   2791 filtered -   60 saved (length * erate cutoff 4170.87)
       12 -   4003 overlaps -   4003 scored -   3943 filtered -   60 saved (length * erate cutoff 3429.43)
       13 -   2533 overlaps -   2533 scored -   2473 filtered -   60 saved (length * erate cutoff 1521.82)
       14 -   3916 overlaps -   3916 scored -   3856 filtered -   60 saved (length * erate cutoff 1823.84)
       15 -     18 overlaps -     18 scored -      0 filtered -   18 saved (no filtering)
       16 -   1591 overlaps -   1591 scored -   1531 filtered -   60 saved (length * erate cutoff 2416.37)
       17 -     42 overlaps -     42 scored -      0 filtered -   42 saved (no filtering)
       18 -    504 overlaps -    504 scored -    444 filtered -   60 saved (length * erate cutoff 2311.73)
       19 -   1176 overlaps -   1176 scored -   1116 filtered -   60 saved (length * erate cutoff 1399.17)
...
```
So any way to speed this process ?

Best,
Ryan

**Comment by brianwalenz** [comment]
No way at the moment.  We want to make some changes in this area in the near future.

**Comment by RyanBio** [comment]
Sorry for late. Very hope to see the changes.
Well, some parallel jobs which are high IO intensive seem to be inefficient on the grid. As i know little about distributed computing, any suggestion about how to optimization jobs for a large genome project ?

**Comment by brianwalenz** [comment]
The only task known to be traumatic to disk is read correction.  See http://canu.readthedocs.io/en/latest/parameter-reference.html?highlight=staging#file-staging for a solution that can copy the data to a local device before computing.

Parallel overlap store building will do a massive amount of sequential reading and writing.  Most filesystems handle this load well.

**Comment by RyanBio** [comment]
Thanks, Brain. I will have a try.

**Comment by skoren** [comment]
The latest code eliminates one pass of the store which was single-threaded and uses information from the store build process instead.

**Comment by GWheelerEB** [comment]
For clarification, does this latest code eliminate the filterCorrectionOverlaps stage altogether? I was having trouble with that step, as by my estimates it was going to run for 5 days (on 1 CPU, 4 GB RAM). Seeing this thread, I updated from 1.6 +13 to 1.6 +56 changes, and when executed, instead of starting that step over, canu skipped straight to corStore construction. Note that I did not manually remove the 2-correction folder or its contents before re-executing.

Is this functioning as intended, or is it moving forward using an only-partially-completed set of globalScores?

**Comment by brianwalenz** [comment]
I think you'll need to backup to release 1.6.  (+13 probably has a bug in it)

The scores are computed and saved when the ovlStore is built.  Without the scores, I think the filter is essentially disabled.  The alternate is to back up and recreate the ovlStore, but you'll probably need to then recreate the input overlaps since they're deleted as soon as the store exists.

So:  the current tip seems to be working well, but you basically need to start the assembly again.  1.6 works, just blow away the 2-correction directory.

How big is the ovlStore?  5 days seems very slow.

**Comment by GWheelerEB** [comment]
My data set is approximately 25x coverage of PacBio data of a 5 gigabase carnivorous plant genome. After the initial short-read filtration, it looks like I have 19344780 reads 122.3 gigabase of sequence in 19.3 million reads. It took the filterCorrectionOverlap step 24 hours to run through about 6 million reads (if I am understanding its log outputs correctly), so that would be 77 hours to run through them all (so not 5 days, but still a pretty long time).

Starting over is not a great option for me, as I've already chewed through a tremendous amount of my compute-time allocation. If rolling back to 1.6 and letting the computation continue for the required three days is the only way to go, then so be it. I'm curious as to what the updated tip is currently doing right now, though, if it should not be able to run without the previously-calculated scores. The only output I have from the active step is: 

> -- BEGIN CORRECTION
> --
> -- Set corMinCoverage=4 based on read coverage of 24.
> -- Global filter scores will be estimated.
> -- Computing correction layouts.
> ----------------------------------------
> -- Starting command on Fri Sep 22 11:36:43 2017 with 192858.955 GB free disk space
> 
>     cd correction
>     /users/PAS1172/osu0330/canu/Linux-amd64/bin/generateCorrectionLayouts \
>       -G ./salata.gkpStore \
>       -O ./salata.ovlStore \
>       -C ./salata.corStore.WORKING \
>       -ec 4 \
>       -eC 80 \
>     > ./salata.corStore.err 2>&1

In 2.5 hours, it has produced one file, seqDB.v001.dat, with a size of 50 megabytes and slowly growing. The file's contents are not readable text, so I have no idea if this is normal or not.

**Comment by brianwalenz** [comment]
The filter is simply reading every overlap (in order) and computing some statistics.  Thus, it needs to read all the data in the ovlStore directory - should be TB's of data there.  I only asked because 5 days seems awfully slow, and this should be limited more by disk throughput than CPU time.

What you have running now will run, but with no overlap scores, correction quality will be worse.  Will it matter?  Don't know.  Assemblies were slightly but noticeably inferior when we had bugs in the filtering or disabled entirely.

The binary file is normal.  I'll add some status reporting.  The long term goal is to make generateCorrectionLayouts also run in parallel, so these multi-day passes through all data will go away.  It doesn't help you any right now; you're stuck doing two passes through the overlaps - one for scores and one for layouts.

**Comment by GWheelerEB** [comment]
Alright, thank you for that clarification. I have rolled back to canu-1.6 and am rerunning starting back from the filterCorrectionOverlaps stage. I am already working with a large and repetitive genome, so I'd like to take advantage of any improvement in assembly I can get, even if slight. Hopefully my time estimates are correct and 

About disk throughput: my data is currently stored in the Ohio Supercomputer's 'scratch' space, which is a high-performance shared area. This is faster than normal user storage but slower than a node's local disk. The ovsStore directory weighs in at 31TB for my data, though, which is far more than a single node's $TMPDIR can handle. However, the section of the user guide linked in your previous comments seems to suggest that setting fistageDirectory=$TMPDIR, and then copying the much smaller gkpStore to there, can greatly accelerate the process. Is that correct?

**Comment by brianwalenz** [comment]
Staging will greatly accelerate the actual generation of corrected reads, and will be done automatically for you if you set the stage options.  For this part, unfortunately, it's just one process needing to slug through 30 TB of data.

Worse, after this 'filter' gets done (creating a globalScores file), 'generateCorrectedReads' will kick off and do _another_ pass through all this data to figure out which reads should be corrected (creating a set of 'readsToCorrect' files).  Once that is done, a set of jobs running 'correctReads.sh' will be launched - it is these jobs that will stage the gkpStore to local disk.

**Comment by GWheelerEB** [comment]
I would like to clarify what stage I am on now and if I'm doing something wrong. The current command is "generateCorrectionLayouts". For 25x coverage of 5 gigabase, it ran for 96 hours (using 1 core, 1.26 GB RAM), created a 68 TB .dat file in salata.corStore.WORKING, and then was killed due to exceeding walltime. I tried to execute this again hoping it would proceed the rest of the way, but it just wiped the file and started over. I have added stageDirectory="\$TMPDIR" to the canu command, but as far as I can tell, it isn't doing anything with that at this phase. Is this working correctly? If so, is there any way I can estimate how long it will take/how large the file will become? If not, is there any recommendation for a fix?

It is hard for me to tell which is the same issue as above -- seems to be a similar single-thread bottleneck. I can create a separate issue thread for this if preferable.

**Comment by dkainer** [comment]
I am having the same issue. I find that setting stageDirectory does nothing. All temporary files seem to be written to the working directory throughout correction. I am using a PBSpro grid but where each node/job has up to 440 Gb of temporary scratch space but it does not seem to be used at all.

Regarding the single threaded generateCorrectionLayouts process - will providing more memory to the executive process speed this up? The default settings gave it just 4 Gb and I could easily run this with way more.

**Comment by skoren** [comment]
As @brianwalenz said above, the staging is only done for computing the consensus of the corrected reads. This is the only step which does significant random seek on disk which benefits from local staging. 

For the overlap filtering, the process streams the entire store which is predictable and easy to optimize from the OS perspective, so most parallel filesystems will handle this type of access very well. The time you'd pay to move the folder to temporary space would far exceed any savings. Memory won't make a difference because it's just streaming, it doesn't actually need more than just the overlaps for any given sequence at a time. It can potentially be parallelized and will be in the future. The latest tip also eliminates one single-threaded pass but that has had limited testing to date so use at your own risk.
